### PR TITLE
Add `User-Agent` header to all outbound requests coming from the API

### LINKED ISSUE #826: User-Agent setting needed for Wikimedia watermarked images
## Sentry link

https://sentry.io/share/issue/699d8b5534e34b03a0dd721e32cbc7b9/

## Description

Wikimedia images requests fail at `v1/images/{identifier}/watermark/` endpoint due to Wikimedia's policies. Sentry shows:

> reason: Forbidden. Please comply with the User-Agent policy: https://meta.wikimedia.org/wiki/User-Agent_policy

So we cannot get the image in the API, and the below function fails; therefore, the request to the endpoint fails (500 error). Sentry shows that this is a new issue.

https://github.com/WordPress/openverse-api/blob/81b7860e2662f4b135919da0728eea194f15c260/api/catalog/api/utils/watermark.py#L140-L150

## Reproduction

See the error at https://api-dev.openverse.engineering/v1/images/f225be68-0b96-465e-8e04-836aae2c1419/watermark/

## Additional context

Related to #483.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #826 by @sarayourfriend 

## Description

The fix should be applied to all outbound requests that could make a request to Wikimedia, not just in the Watermark endpoint. This is _all_ of our outbound requests, so I've had to update all our outbound requests for it.

I tried to explore some ways of centralising this configuration, primarily by subclasses `requests.Session` and forcing a `purpose` kwarg to be present on all calls to `::request` that would then configure the UA header, but it seemed heavy and I wasn't sure exactly how it would work or if folks would like that. I think it would make things a little easier to manage, or at least easier to not make mistakes with. But we also use `grequests` and `urlopen` which would still need their own bespoke solution... so I opted for the simplest approach across the board for now. We can improve this/generalise/abstract on it later if we want.

## Testing Instructions

Run the app locally and ensure there are no issues.

Check out the unit tests. The ones involving urlopen were a complete nightmare to write!

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ services:
     environment:
       PORT: 8222
       MALLOC_ARENA_MAX: 2
-    command: ["-enable-url-source"]
+    command: ["-enable-url-source -forward-headers User-Agent"]
```
Comment (sarayourfriend) [inline docker-compose.yml:21]: This new option will need to be ported up to the infrastructure repository as well, if it is merged.

**Review by dhruvkb — APPROVED** [review]
I think this is a reasonable configuration for now. The new env var you've added `OUTBOUND_USER_AGENT_TEMPLATE` should also be added to the `env.template` file (if only as a comment because the default value is good enough).

I agree that in the future we can use a central solution to manage requests but that will require all HTTP requests to first use the same libraries which will take a significant amount of time.

**Review by AetherUnbound — APPROVED** [review]
This is great, thank you for adding thorough tests too. Agreed that a more centralized implementation would be cool but the variety of request methods we have precludes that. I think the "purpose" interpolation you've added is excellent 💯 

I have a few comments around the tests but the operational code is ✔️ IMO

File: api/test/unit/utils/watermark_test.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+from pathlib import Path
+from typing import Callable
+
+import pytest
+from requests import Request, Response
+
+from catalog.api.utils.watermark import HEADERS, watermark
+
+
+_MOCK_IMAGE_PATH = Path(__file__).parent / ".." / ".." / "factory"
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply api/test/unit/utils/watermark_test.py:12]: I feel that this syntax is rather obscure compared to regular directory navigation.

File: api/test/unit/views/image_views_test.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def _default_response_factory(req: Request) -> Response:
+        res = Response()
+        res.url = req.url
+        res.status_code = 200
+        res._content = _MOCK_IMAGE_BYTES
+        return res
+
+
+@pytest.fixture(autouse=True)
+def requests(monkeypatch) -> RequestsFixture:
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply api/test/unit/views/image_views_test.py:42]: I would rather do this refactor in another PR as part of another issue that more careful looks at the state of our unit test fixtures in general and thoughtfully establishes a pattern. For now most of these have subtle differences that would be tedious to abstract.

**Comment by sarayourfriend** [comment]
Blocked by https://github.com/WordPress/openverse-api/pull/875

File: api/test/unit/views/image_views_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply api/test/unit/views/image_views_test.py:42]: Sounds good 🙂

**Review by AetherUnbound — APPROVED** [review]
New changes look good 👍🏼

**Pull request merged**